src/distributions/uniform.rs

```diff
@@ -534,59 +533,110 @@ macro_rules! uniform_float_impl {
-                let scale = high - low;
-                let offset = low - scale;
-                UniformFloat {
-                    scale: scale,
-                    offset: offset,
+                assert!(low.is_finite() && high.is_finite(), "Uniform::new called with non-finite boundaries");
+
+                let mut scale = high - low;
+
+                let max_rand = (::core::$uty::MAX >> $bits_to_discard)
+                               .into_float_with_exponent(0) - 1.0;
+
+                while scale * max_rand + low >= high {
+                    scale = <$ty>::from_bits(scale.to_bits() - 1);
                 }
+
+                debug_assert!(scale >= 0.0);
+
+                UniformFloat { low, scale }
             }
 
             fn new_inclusive(low: Self::X, high: Self::X) -> Self {
-                assert!(low <= high,
-                        "Uniform::new_inclusive called with `low > high`");
-                let scale = high - low;
-                let offset = low - scale;
-                UniformFloat {
-                    scale: scale,
-                    offset: offset,
+                assert!(low <= high, "Uniform::new_inclusive called with `low > high`");
+                assert!(low.is_finite() && high.is_finite(), "Uniform::new_inclusive called with non-finite boundaries");
+
+                let mut scale = high - low;
+
+                let max_rand = (::core::$uty::MAX >> $bits_to_discard)
+                               .into_float_with_exponent(0) - 1.0;
+
+                if scale * max_rand + low > high {
+                    // Adjust scale lower until the max value we can generate is `high`
+                    while scale * max_rand + low > high {
+                        scale = <$ty>::from_bits(scale.to_bits() - 1);
+                    }
+                } else if high != low {
+                    // No need to loop here since even if `scale` is strictly not as big
+                    // as it could be, we'll still generate good values.
+                    // But give it one try to increase our chances that
+                    // Uniform::new_inclusive is actually inclusive of the high value.
+                    let new_scale = <$ty>::from_bits(scale.to_bits() + 1);
+                    if new_scale * max_rand + low <= high {
+                        scale = new_scale;
+                    }
                 }
+
+                debug_assert!(scale >= 0.0);
+
+                UniformFloat { low, scale }
             }
 
             fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Self::X {
                 // Generate a value in the range [1, 2)
                 let value1_2 = (rng.$next_u() >> $bits_to_discard)
                                .into_float_with_exponent(0);
+
+                // Get a value in the range [0, 1) in order to avoid
+                // overflowing into infinity when multiplying with scale
+                let value0_1 = value1_2 - 1.0;
```

Comment:
This loses a bit of precision everywhere just to handle scales close to what would round to infinity. I'm not sure it's worth it; we could instead assert that `scale * 2` is finite in the constructor.

Reply:
It shouldn't lose any precision. Subtracting 1.0 should not result in any rounding or loss of data. In fact, it looks to me like it should gain a tiny amount of precision since we're directly adding `low` rather than the calculated (and thus possibly slightly rounded) `offset`.

But it does lose performance.

Either way, not having this here is certainly possible, but I think it'll require a decent amount more code in the constructor. We'll have to worry about not just adjusting the `scale`, but also adjusting `offset` such that rounding can't result in values lower than `low`.

So it's a tradeoff of slower and more code complexity in the constructor, vs. saving around 8% in the sampling code.

Reply:
Sorry, you're right about the precision.

I don't know what's the right answer regarding performance.